x-pack/plugins/endpoint/public/applications/endpoint/types.ts

```diff
@@ -8,18 +8,31 @@ import { Dispatch, MiddlewareAPI } from 'redux';
 import { CoreStart } from 'kibana/public';
 import { EndpointListState } from './store/endpoint_list';
 import { AppAction } from './store/action';
-import { AlertResultList } from '../../../common/types';
+import { AlertResultList, Immutable } from '../../../common/types';
 
-export type MiddlewareFactory = (
+export type MiddlewareFactory<State = GlobalState> = (
   coreStart: CoreStart
-) => (
-  api: MiddlewareAPI<Dispatch<AppAction>, GlobalState>
-) => (next: Dispatch<AppAction>) => (action: AppAction) => unknown;
+) => {
+  middleware: (
+    api: MiddlewareAPI<Dispatch<AppAction>, State>
+  ) => (next: Dispatch<AppAction>) => (action: AppAction) => unknown;
+  start?: () => void;
+};
 
 export interface GlobalState {
   readonly endpointList: EndpointListState;
   readonly alertList: AlertListState;
 }
 
+export interface EndpointAppLocation {
```

Comment:
(optional) do we need our own type or could we use `RouteComponentProps` from `react-router-dom` (as in `RouteComponentProps['location']` )

Reply:
react router dom also uses the `Location` type from History and since we have a state that is going to be never, i think it'd be more accessibly readable to use our own type